File: src/main/java/com/ftn/ues/email_client/dao/elastic_search/Attachment.java
package com.ftn.ues.email_client.dao.elastic_search;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.ftn.ues.email_client.dao.DirectMapping;
import com.ftn.ues.email_client.model.AttachmentRaw;
import com.ftn.ues.email_client.model.Message;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.experimental.SuperBuilder;
import org.javatuples.Pair;
import org.joda.time.DateTime;
import org.springframework.data.annotation.Id;
import org.springframework.data.elasticsearch.annotations.Document;
import org.springframework.data.elasticsearch.annotations.Field;
import org.springframework.data.elasticsearch.annotations.FieldType;

import java.util.Arrays;
import java.util.Base64;

@Data
@AllArgsConstructor
@SuperBuilder
@Document(indexName = "attachment", shards = 2)
public class Attachment extends DirectMapping<com.ftn.ues.email_client.model.Attachment> {

    @Id
    private Long id;

    @Field(type = FieldType.Text)
    private String path;

    @Field(type = FieldType.Text)
    private String mimeType;

    @Field(type = FieldType.Text)
    private String name;

    @Field(type = FieldType.Long)
    private Long message;

    @Field(type = FieldType.Text)
    private String data;

    @Field(type = FieldType.Nested)
    private AttachmentMetadata metadata;

    @Builder.Default
    private Boolean deleted = false;

    public Attachment(Pair<com.ftn.ues.email_client.model.Attachment, AttachmentRaw> attachmentRawPair) {
        this(attachmentRawPair.getValue0());
        this.data = Base64.getEncoder().encodeToString(attachmentRawPair.getValue1().getData());
    }

    public Attachment(com.ftn.ues.email_client.model.Attachment object) {
        super(object);
        this.id = object.getId();
        this.path = object.getPath();
        this.mimeType = object.getMimeType();
        this.name = object.getName();
        this.message = object.getMessage().getId();
        this.deleted = object.getDeleted();
    }

    @Override
    public com.ftn.ues.email_client.model.Attachment getModelObject() {
        return com.ftn.ues.email_client.model.Attachment.builder()
                .id(id)
                .path(path)
                .mimeType(mimeType)
                .name(name)
                .message(Message.builder().id(message).build())
                .deleted(deleted)
                .build();
    }

    @Data
    @AllArgsConstructor
    public class AttachmentMetadata {

        @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss:SSSZZ")
        private DateTime date;

        @Field(name = "content_type")
        private String contentType;

        private String language;
        private String content;

        @Field(name = "content_length")
        private Long contentLength;

    }
}

File: src/main/scala/com/ftn/ues/email_client/dao/elastic/Attachment.scala
package com.ftn.ues.email_client.dao.elastic

import play.api.libs.json.{Json, OFormat}


case class AttachmentMetadata(
                               date: Option[String],
                               language: Option[String],
                               content: Option[String]
                             )
object AttachmentMetadata{
  implicit val format: OFormat[AttachmentMetadata] = Json.format[AttachmentMetadata]
}

case class Attachment(
                       id: Long,
                       path: String,
                       mimeType: String,
                       name: String,
                       messageId: Long,
                       data: Option[String]=None,
                       metadata: Option[AttachmentMetadata],
                       deleted: Boolean
                     )

object Attachment {
  def apply(attachment: com.ftn.ues.email_client.model.Attachment, data: String): Attachment = {
    Attachment(
      attachment.getId,
      attachment.getPath,
      attachment.getMimeType,
      attachment.getName,
      attachment.getMessage.getId,
      Some(data),
      None,
      attachment.getDeleted
    )
  }

  implicit val format: OFormat[Attachment] = Json.format[Attachment]
}

